packages/adapter/lib/adapter/adapter.js

```diff
@@ -8229,93 +8228,105 @@ function Adapter(options) {
-                this.log.info('unsubscrubeForeignStates not processed because States database not connected');
-                return tools.maybeCallbackWithError(callback, tools.ERRORS.ERROR_DB_CLOSED);
-            }
-
             // Todo check rights for options
             if (typeof options === 'function') {
                 callback = options;
                 options = null;
             }
 
+            if (pattern instanceof RegExp) {
+                return tools.maybeCallbackWithError(
+                    callback,
+                    `Regexp is not supported for "unsubscribeForeignStates", received "${pattern.toString()}"`
+                );
+            }
+
+            if (!adapterStates) {
+                // if states is no longer existing, we do not need to unsubscribe
+                this.log.info('unsubscrubeForeignStates not processed because States database not connected');
+                return tools.maybeCallbackWithError(callback, tools.ERRORS.ERROR_DB_CLOSED);
+            }
+
             if (this.autoSubscribe && typeof pattern === 'string') {
-                for (let s = 0; s < this.autoSubscribe.length; s++) {
-                    if (
-                        pattern === '*' ||
-                        pattern.substring(0, this.autoSubscribe[s].length + 1) === `${this.autoSubscribe[s]}.`
-                    ) {
+                for (const autoSub of this.autoSubscribe) {
+                    if (pattern === '*' || pattern.substring(0, autoSub.length + 1) === `${autoSub}.`) {
                         // remove this pattern from adapter list
-                        adapterStates.getState(`system.adapter.${this.autoSubscribe[s]}.subscribes`, (err, state) => {
-                            if (!state || !state.val) {
-                                return;
-                            }
-                            let subs;
-                            try {
-                                subs = JSON.parse(state.val);
-                            } catch {
-                                logger.error(
-                                    `${this.namespaceLog} Cannot parse subscribes for "${this.autoSubscribe[s]}.subscribes"`
-                                );
-                                return;
-                            }
-                            if (!subs[pattern]) {
-                                return;
-                            }
-                            if (subs[pattern][this.namespace] === undefined) {
-                                return;
-                            }
+                        let state;
+                        try {
+                            state = await adapterStates.getState(`system.adapter.${autoSub}.subscribes`);
+                        } catch {
+                            // ignore
+                        }
+                        if (!state || !state.val) {
+                            return;
```

Comment:
ppuuh ... really return=? We are ina aoop ...hm ... continue better? (so in general=

Reply:
probably better, also strange to return w/o calling cb

Reply:
probably for all returns in this loop